akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala

```diff
@@ -0,0 +1,165 @@
+      val accumulated = collection.mutable.Buffer.empty[In]
+
+      private var subSource = OptionVal.none[SubSourceOutlet[In]]
+      private var subSink = OptionVal.none[SubSinkInlet[Out]]
+
+      private var downstreamCause = OptionVal.none[Throwable]
+
+      setHandlers(in, out, this)
+
+      override def postStop(): Unit = {
+        //this covers the case when the nested flow was never materialized
+        matPromise.tryFailure(new AbruptStageTerminationException(this))
+        super.postStop()
+      }
+
+      override def onPush(): Unit = {
+        if (subSource.isDefined) {
+          subSource.get.push(grab(in))
+        } else {
+          accumulated.append(grab(in))
+          if (accumulated.size == n) {
+            materializeFlow()
+          } else {
+            //gi'me some more!
+            pull(in)
+          }
+        }
+      }
+
+      override def onUpstreamFinish(): Unit = {
+        if (subSource.isDefined) {
+          subSource.get.complete()
+        } else {
+          materializeFlow()
+        }
+      }
+
+      override def onUpstreamFailure(ex: Throwable): Unit = {
+        if (subSource.isDefined) {
+          subSource.get.fail(ex)
+        } else {
+          //flow won't be materialized, so we have to complete the future with a failure indicating this
+          matPromise.failure(new NeverMaterializedException(ex))
+          super.onUpstreamFailure(ex)
+        }
+      }
+
+      override def onPull(): Unit = {
+        if (subSink.isDefined) {
+          //delegate to subSink
+          subSink.get.pull()
+        } else if (accumulated.size < n) {
+          pull(in)
+        } else if (accumulated.size == n) { //corner case for n = 0, can be handled in FlowOps
+          materializeFlow()
+        }
+      }
+
+      override def onDownstreamFinish(cause: Throwable): Unit = {
+        if (subSink.isEmpty) {
+          downstreamCause = OptionVal.Some(cause)
+        } else {
+          subSink.get.cancel(cause)
+        }
+      }
+
+      def materializeFlow(): Unit =
+        try {
+          val prefix = accumulated.toVector
+          accumulated.clear()
+          subSource = OptionVal.Some(new SubSourceOutlet[In]("subSource"))
+          subSource.get.setHandler {
+            new OutHandler {
+              override def onPull(): Unit = {
+                if (!isClosed(in) && !hasBeenPulled(in)) {
+                  pull(in)
+                }
+              }
+
+              override def onDownstreamFinish(cause: Throwable): Unit = {
+                if (!isClosed(in)) {
+                  cancel(in, cause)
+                }
+                //super.onDownstreamFinish(cause)
+              }
+            }
+          }
+          subSink = OptionVal.Some(new SubSinkInlet[Out]("subSink"))
+          subSink.get.setHandler {
+            new InHandler {
+              override def onPush(): Unit = {
+                push(out, subSink.get.grab())
```

Comment:
This is in the fast path, so maybe get rid of the `.get` which introduces an extra null check. If you do that also get rid of all the other `.get` in here.

Reply:
This particular one hasn't been fixed yet ;)